pycrostates/cluster/aahc.py

```diff
@@ -0,0 +1,386 @@
+        data: NDArray[float],
+        n_clusters: int,
+        ignore_polarity: bool,
+        normalize_input: bool,
+    ) -> Tuple[float, NDArray[float], NDArray[int]]:
+        """Run the AAHC algorithm."""
+        gfp_sum_sq = np.sum(data**2)
+        maps, segmentation = AAHCluster._compute_maps(
+            data, n_clusters, ignore_polarity, normalize_input
+        )
+        map_corr = _corr_vectors(data, maps[segmentation].T)
+        gev = np.sum((data * map_corr) ** 2) / gfp_sum_sq
+        return gev, maps, segmentation
+
+    @staticmethod
+    def _compute_maps(
+        data: NDArray[float],
+        n_clusters: int,
+        ignore_polarity: bool,
+        normalize_input: bool,
+    ) -> Tuple[NDArray[float], NDArray[int]]:
+        """Compute microstates maps."""
+        n_chan, n_frame = data.shape
+
+        cluster = data.copy()
+        cluster /= np.linalg.norm(cluster, axis=0, keepdims=True)
+
+        if normalize_input:
+            data = cluster.copy()
+
+        GEV = np.sum(data * cluster, axis=0)
+
+        assignment = np.arange(n_frame)
+
+        while cluster.shape[1] > n_clusters:
+
+            to_remove = np.argmin(GEV)
+            orphans = assignment == to_remove
+
+            old_cluster = cluster[:, to_remove].copy()
+
+            cluster = np.delete(cluster, to_remove, axis=1)
+            GEV = np.delete(GEV, to_remove, axis=0)
+            assignment[assignment > to_remove] = (
+                assignment[assignment > to_remove] - 1
+            )
+
+            fit = data[:, orphans].T @ cluster
+            if ignore_polarity:
+                fit = np.abs(fit)
+            new_assignment = np.argmax(fit, axis=1)
+            assignment[orphans] = new_assignment
+
+            cluster_to_update = np.unique(new_assignment)
+            for c in cluster_to_update:
+                members = assignment == c
+                if ignore_polarity:
+
+                    old_weight = len(new_assignment == c) / members.sum()
+
+                    sgn = np.sign(old_cluster @ cluster[:, c])
+
+                    v0 = (
+                        old_weight * sgn * old_cluster
+                        + (1.0 - old_weight) * cluster[:, c]
+                    )
+
+                    v, _, converged = AAHCluster._first_principal_component(
+                        data[:, members], v0, max_iter=n_chan
+                    )
+                    if not converged:
+                        # fall back to covariance estimation
+                        # and eigenvalue computation
+                        Cxx = data[:, members] @ data[:, members].T
+                        _, V = np.linalg.eigh(Cxx)
+                        v = V[:, -1]
+
+                    cluster[:, c] = v
+                else:
+                    cluster[:, c] = np.mean(data[:, members], axis=1)
+                    cluster[:, c] /= np.linalg.norm(
+                        cluster[:, c], axis=0, keepdims=True
+                    )
```

Comment:
One more question about this part, I'm a bit surprised by the difference between the steps when `ignore_polarity` is set to True and the steps when it is set to False. 
Could you expand a bit on those, either here, or even maybe in a couple of comments in the code?

Mostly, I'm trying to build confidence that it shouldn't be:

```
if ignore_polarity:
    # do some extra-steps

cluster[:, c] = np.mean(data[:, members], axis=1)
cluster[:, c] /= np.linalg.norm(cluster[:, c], axis=0, keepdims=True)
```

Also, do you know if there is a scikit-learn implementation of this algorithm (which would be the case `ignore_polarity=False`)? I found `sklearn.cluster.AgglomerativeClustering` but I'm not sure it's the same algo.

Reply:
Nevermind, @vferat convinced me.